codegen: redo metadata

ahjohannessen opened the pull request:
- change code generator to parameterize metadata instead
   of hardcoding to `io.grpc.Metadata`.

 - add `client` and `service` methods to service companion
   that take a function from `A => Metadata` and
   `Metadata => Either[String, A]`, respectively.

 - add option to specify suffix for generated code.

 - change `client` method to take function
   `c: CallOptions => CallOptions` in order
   to be able to use `withDeadline` for each
   call. Current approach requires newing up
   a new client for each call as a deadline is
   absolute.

ahjohannessen commented:
@rossabaker @fiadliel I started out with `Kleisli`, but it turned out a bit messy due to this scenario:
`_root_.fs2.Stream[F, Result]`.

This is somewhat what it was before, but now one can provide `A => Metadata` for stub and `Metadata => A` for service.

ahjohannessen commented:
The following protobuf:
```
syntax = "proto2";
import "scalapb/scalapb.proto";

package ss.be.rpc.protos;

message HelloRequest {
  required string name = 1;
}

message HelloReply {
  required string message = 1;
}
```
and
```
syntax = "proto3";

import "scalapb/scalapb.proto";
import "hello.proto";

package ss.be.rpc.protos;

service Greeter {
  rpc SayHelloV1 (HelloRequest) returns (HelloReply) {}
  rpc SayHelloV2 (HelloRequest) returns (stream HelloReply) {}
  rpc SayHelloV3 (stream HelloRequest) returns (HelloReply) {}
  rpc SayHelloV4 (stream HelloRequest) returns (stream HelloReply) {}
}
```

results in this service when using empty string as service suffix:
```scala
package ss.be.rpc.protos

import _root_.cats.implicits._

trait Greeter[F[_], A] {
  def sayHelloV1(request: ss.be.rpc.protos.HelloRequest, ctx: A): F[ss.be.rpc.protos.HelloReply]
  def sayHelloV2(request: ss.be.rpc.protos.HelloRequest, ctx: A): _root_.fs2.Stream[F, ss.be.rpc.protos.HelloReply]
  def sayHelloV3(request: _root_.fs2.Stream[F, ss.be.rpc.protos.HelloRequest], ctx: A): F[ss.be.rpc.protos.HelloReply]
  def sayHelloV4(request: _root_.fs2.Stream[F, ss.be.rpc.protos.HelloRequest], ctx: A): _root_.fs2.Stream[F, ss.be.rpc.protos.HelloReply]
}
object Greeter {
  
  type GreeterMeta[F[_]] = Greeter[F, _root_.io.grpc.Metadata]
  
  def stub[F[_]: _root_.cats.effect.ConcurrentEffect](channel: _root_.io.grpc.Channel, callOptions: _root_.io.grpc.CallOptions = _root_.io.grpc.CallOptions.DEFAULT): GreeterMeta[F] = {
    client[F, _root_.io.grpc.Metadata](channel, identity, callOptions)
  }
  
  def bindService[F[_]: _root_.cats.effect.ConcurrentEffect](serviceImpl: GreeterMeta[F]): _root_.io.grpc.ServerServiceDefinition = {
    service[F, _root_.io.grpc.Metadata](serviceImpl, identity)
  }
  
  def client[F[_]: _root_.cats.effect.ConcurrentEffect, A](channel: _root_.io.grpc.Channel, f: A => _root_.io.grpc.Metadata, callOptions: _root_.io.grpc.CallOptions = _root_.io.grpc.CallOptions.DEFAULT): Greeter[F, A] = new Greeter[F, A] {
    def sayHelloV1(request: ss.be.rpc.protos.HelloRequest, ctx: A): F[ss.be.rpc.protos.HelloReply] = {
      _root_.org.lyranthe.fs2_grpc.java_runtime.client.Fs2ClientCall[F](channel, _root_.ss.be.rpc.protos.GreeterGrpc.METHOD_SAY_HELLO_V1, callOptions).flatMap(_.unaryToUnaryCall(request, f(ctx)))
    }
    def sayHelloV2(request: ss.be.rpc.protos.HelloRequest, ctx: A): _root_.fs2.Stream[F, ss.be.rpc.protos.HelloReply] = {
      _root_.fs2.Stream.eval(_root_.org.lyranthe.fs2_grpc.java_runtime.client.Fs2ClientCall[F](channel, _root_.ss.be.rpc.protos.GreeterGrpc.METHOD_SAY_HELLO_V2, callOptions)).flatMap(_.unaryToStreamingCall(request, f(ctx)))
    }
    def sayHelloV3(request: _root_.fs2.Stream[F, ss.be.rpc.protos.HelloRequest], ctx: A): F[ss.be.rpc.protos.HelloReply] = {
      _root_.org.lyranthe.fs2_grpc.java_runtime.client.Fs2ClientCall[F](channel, _root_.ss.be.rpc.protos.GreeterGrpc.METHOD_SAY_HELLO_V3, callOptions).flatMap(_.streamingToUnaryCall(request, f(ctx)))
    }
    def sayHelloV4(request: _root_.fs2.Stream[F, ss.be.rpc.protos.HelloRequest], ctx: A): _root_.fs2.Stream[F, ss.be.rpc.protos.HelloReply] = {
      _root_.fs2.Stream.eval(_root_.org.lyranthe.fs2_grpc.java_runtime.client.Fs2ClientCall[F](channel, _root_.ss.be.rpc.protos.GreeterGrpc.METHOD_SAY_HELLO_V4, callOptions)).flatMap(_.streamingToStreamingCall(request, f(ctx)))
    }
  }
  
  def service[F[_]: _root_.cats.effect.ConcurrentEffect, A](serviceImpl: Greeter[F, A], f: _root_.io.grpc.Metadata ⇒ A): _root_.io.grpc.ServerServiceDefinition = {
    _root_.io.grpc.ServerServiceDefinition
      .builder(_root_.ss.be.rpc.protos.GreeterGrpc.SERVICE)
      .addMethod(_root_.ss.be.rpc.protos.GreeterGrpc.METHOD_SAY_HELLO_V1, _root_.org.lyranthe.fs2_grpc.java_runtime.server.Fs2ServerCallHandler[F].unaryToUnaryCall[ss.be.rpc.protos.HelloRequest, ss.be.rpc.protos.HelloReply]((r, m) ⇒ serviceImpl.sayHelloV1(r, f(m))))
      .addMethod(_root_.ss.be.rpc.protos.GreeterGrpc.METHOD_SAY_HELLO_V2, _root_.org.lyranthe.fs2_grpc.java_runtime.server.Fs2ServerCallHandler[F].unaryToStreamingCall[ss.be.rpc.protos.HelloRequest, ss.be.rpc.protos.HelloReply]((r, m) ⇒ serviceImpl.sayHelloV2(r, f(m))))
      .addMethod(_root_.ss.be.rpc.protos.GreeterGrpc.METHOD_SAY_HELLO_V3, _root_.org.lyranthe.fs2_grpc.java_runtime.server.Fs2ServerCallHandler[F].streamingToUnaryCall[ss.be.rpc.protos.HelloRequest, ss.be.rpc.protos.HelloReply]((r, m) ⇒ serviceImpl.sayHelloV3(r, f(m))))
      .addMethod(_root_.ss.be.rpc.protos.GreeterGrpc.METHOD_SAY_HELLO_V4, _root_.org.lyranthe.fs2_grpc.java_runtime.server.Fs2ServerCallHandler[F].streamingToStreamingCall[ss.be.rpc.protos.HelloRequest, ss.be.rpc.protos.HelloReply]((r, m) ⇒ serviceImpl.sayHelloV4(r, f(m))))
      .build()
  }
}
```

rossabaker commented on sbt-java-gen/src/main/scala/Fs2GrpcServicePrinter.scala:
```diff
@@ -73,28 +78,74 @@ class Fs2GrpcServicePrinter(service: ServiceDescriptor, di: DescriptorImplicits)
   private[this] def serviceBinding: PrinterEndo = {
     _.add(
-      s"def bindService[F[_]: $ConcurrentEffect](serviceImpl: ${service.name}Fs2Grpc[F]): $ServerServiceDefinition = {").indent
+      s"def service[F[_]: $ConcurrentEffect, $Ctx](serviceImpl: $serviceNameFs2[F, $Ctx], f: $Metadata ⇒ $Ctx): $ServerServiceDefinition = {")
```
Unicode arrows are deprecated in Scala 2.13.  I didn't scan the review for others:

```suggestion
      s"def service[F[_]: $ConcurrentEffect, $Ctx](serviceImpl: $serviceNameFs2[F, $Ctx], f: $Metadata => $Ctx): $ServerServiceDefinition = {")
```

rossabaker commented on sbt-java-gen/src/main/scala/Fs2GrpcServicePrinter.scala:
```diff
@@ -73,28 +78,74 @@ class Fs2GrpcServicePrinter(service: ServiceDescriptor, di: DescriptorImplicits)
+
+  private[this] def serviceNameMeta = s"${serviceNameFs2}Meta"
+
+  private[this] def serviceTraitMeta: PrinterEndo =
```
I wonder if this helps upgrade all that much, since it's a new name.  I lean toward not including this, but don't feel strongly.

ahjohannessen replied:
I do not feel strongly by this, the type alias can be removed. Just worried a bit about existing users might be put of by two type params.

rossabaker replied:
It's easier to take things away than add them.  We can bring it back later if it causes confusion.

ahjohannessen commented on sbt-java-gen/src/main/scala/Fs2GrpcServicePrinter.scala:
```diff
@@ -73,28 +78,74 @@ class Fs2GrpcServicePrinter(service: ServiceDescriptor, di: DescriptorImplicits)
 
   private[this] def serviceTrait: PrinterEndo =
-    _.add(s"trait ${serviceName}Fs2Grpc[F[_]] {").indent.call(serviceMethods).outdent.add("}")
+    _.add(s"trait $serviceNameFs2[F[_], $Ctx] {").
```
The motivation is that I do not like having `Fs2Grpc` suffixed in all generated traits. However, default suffix is `Fs2Grpc` so things do not change. Only change is that the generated service file is also suffixed like that instead of `xGrpcFs2.scala`

rossabaker replied:
Oh, I see now.  The "Fs2" was literal in the old version, but is part of the variable name now.  I was having trouble mentally correlating this with the output example and the new `serviceSuffix` setting.

rossabaker replied:
My confusion was from `Fs2` moving from literal to part of the variable name.  I couldn't correlate this with the output example before.  I understand and agree now.

ahjohannessen commented on sbt-java-gen/src/main/scala/Fs2GrpcServicePrinter.scala:
```diff
@@ -47,17 +48,21 @@ class Fs2GrpcServicePrinter(service: ServiceDescriptor, di: DescriptorImplicits)
+    val handler     = s"$Fs2ServerCallHandler[F].${handleMethod(method)}[$inType, $outType]"
+    val serviceCall = s"serviceImpl.${method.name}"
+
+    p.add(s".addMethod($descriptor, $handler((r, m) ⇒ $serviceCall(r, f(m))))")
```
```suggestion
    p.add(s".addMethod($descriptor, $handler((r, m) => $serviceCall(r, f(m))))")
```

ahjohannessen commented:
@rossabaker I have updated the code according to your feedback, removed the type alias - I reckon users themselves can define that type alias if needed. Now the default generated code from `hello_service` is this:

```scala
package ss.be.rpc.protos

import _root_.cats.implicits._

trait GreeterFs2Grpc[F[_], A] {
  def sayHelloV1(request: ss.be.rpc.protos.HelloRequest, ctx: A): F[ss.be.rpc.protos.HelloReply]
  def sayHelloV2(request: ss.be.rpc.protos.HelloRequest, ctx: A): _root_.fs2.Stream[F, ss.be.rpc.protos.HelloReply]
  def sayHelloV3(request: _root_.fs2.Stream[F, ss.be.rpc.protos.HelloRequest], ctx: A): F[ss.be.rpc.protos.HelloReply]
  def sayHelloV4(request: _root_.fs2.Stream[F, ss.be.rpc.protos.HelloRequest], ctx: A): _root_.fs2.Stream[F, ss.be.rpc.protos.HelloReply]
}

object GreeterFs2Grpc {
  
  def stub[F[_]: _root_.cats.effect.ConcurrentEffect](channel: _root_.io.grpc.Channel, callOptions: _root_.io.grpc.CallOptions = _root_.io.grpc.CallOptions.DEFAULT): GreeterFs2Grpc[F, _root_.io.grpc.Metadata] = {
    client[F, _root_.io.grpc.Metadata](channel, identity, callOptions)
  }
  
  def client[F[_]: _root_.cats.effect.ConcurrentEffect, A](channel: _root_.io.grpc.Channel, f: A => _root_.io.grpc.Metadata, callOptions: _root_.io.grpc.CallOptions = _root_.io.grpc.CallOptions.DEFAULT): GreeterFs2Grpc[F, A] = new GreeterFs2Grpc[F, A] {
    def sayHelloV1(request: ss.be.rpc.protos.HelloRequest, ctx: A): F[ss.be.rpc.protos.HelloReply] = {
      _root_.org.lyranthe.fs2_grpc.java_runtime.client.Fs2ClientCall[F](channel, _root_.ss.be.rpc.protos.GreeterGrpc.METHOD_SAY_HELLO_V1, callOptions).flatMap(_.unaryToUnaryCall(request, f(ctx)))
    }
    def sayHelloV2(request: ss.be.rpc.protos.HelloRequest, ctx: A): _root_.fs2.Stream[F, ss.be.rpc.protos.HelloReply] = {
      _root_.fs2.Stream.eval(_root_.org.lyranthe.fs2_grpc.java_runtime.client.Fs2ClientCall[F](channel, _root_.ss.be.rpc.protos.GreeterGrpc.METHOD_SAY_HELLO_V2, callOptions)).flatMap(_.unaryToStreamingCall(request, f(ctx)))
    }
    def sayHelloV3(request: _root_.fs2.Stream[F, ss.be.rpc.protos.HelloRequest], ctx: A): F[ss.be.rpc.protos.HelloReply] = {
      _root_.org.lyranthe.fs2_grpc.java_runtime.client.Fs2ClientCall[F](channel, _root_.ss.be.rpc.protos.GreeterGrpc.METHOD_SAY_HELLO_V3, callOptions).flatMap(_.streamingToUnaryCall(request, f(ctx)))
    }
    def sayHelloV4(request: _root_.fs2.Stream[F, ss.be.rpc.protos.HelloRequest], ctx: A): _root_.fs2.Stream[F, ss.be.rpc.protos.HelloReply] = {
      _root_.fs2.Stream.eval(_root_.org.lyranthe.fs2_grpc.java_runtime.client.Fs2ClientCall[F](channel, _root_.ss.be.rpc.protos.GreeterGrpc.METHOD_SAY_HELLO_V4, callOptions)).flatMap(_.streamingToStreamingCall(request, f(ctx)))
    }
  }
  
  def bindService[F[_]: _root_.cats.effect.ConcurrentEffect](serviceImpl: GreeterFs2Grpc[F, _root_.io.grpc.Metadata]): _root_.io.grpc.ServerServiceDefinition = {
    service[F, _root_.io.grpc.Metadata](serviceImpl, identity)
  }
  
  def service[F[_]: _root_.cats.effect.ConcurrentEffect, A](serviceImpl: GreeterFs2Grpc[F, A], f: _root_.io.grpc.Metadata => A): _root_.io.grpc.ServerServiceDefinition = {
    _root_.io.grpc.ServerServiceDefinition
      .builder(_root_.ss.be.rpc.protos.GreeterGrpc.SERVICE)
      .addMethod(_root_.ss.be.rpc.protos.GreeterGrpc.METHOD_SAY_HELLO_V1, _root_.org.lyranthe.fs2_grpc.java_runtime.server.Fs2ServerCallHandler[F].unaryToUnaryCall[ss.be.rpc.protos.HelloRequest, ss.be.rpc.protos.HelloReply]((r, m) => serviceImpl.sayHelloV1(r, f(m))))
      .addMethod(_root_.ss.be.rpc.protos.GreeterGrpc.METHOD_SAY_HELLO_V2, _root_.org.lyranthe.fs2_grpc.java_runtime.server.Fs2ServerCallHandler[F].unaryToStreamingCall[ss.be.rpc.protos.HelloRequest, ss.be.rpc.protos.HelloReply]((r, m) => serviceImpl.sayHelloV2(r, f(m))))
      .addMethod(_root_.ss.be.rpc.protos.GreeterGrpc.METHOD_SAY_HELLO_V3, _root_.org.lyranthe.fs2_grpc.java_runtime.server.Fs2ServerCallHandler[F].streamingToUnaryCall[ss.be.rpc.protos.HelloRequest, ss.be.rpc.protos.HelloReply]((r, m) => serviceImpl.sayHelloV3(r, f(m))))
      .addMethod(_root_.ss.be.rpc.protos.GreeterGrpc.METHOD_SAY_HELLO_V4, _root_.org.lyranthe.fs2_grpc.java_runtime.server.Fs2ServerCallHandler[F].streamingToStreamingCall[ss.be.rpc.protos.HelloRequest, ss.be.rpc.protos.HelloReply]((r, m) => serviceImpl.sayHelloV4(r, f(m))))
      .build()
  }
}
```

ahjohannessen commented:
Perhaps changing this `Metadata => A` to `Metadata => F[A]` is better, as unmarshalling might fail.

ahjohannessen commented:
@rossabaker I did an additional commit that changes `client` method to take function `c: CallOptions => CallOptions` in order to be able to use `withDeadline` for each call. Current approach requires newing up a new client for each call as a deadline is absolute. I did not change the `stub` method, but that can be changed if needed.

ahjohannessen commented:
`Either[String, A]` could be used to do something like this `err => Status.FAILED_PRECONDITION.withDescription(err).asRuntimeException)`

ahjohannessen commented:
@rossabaker I have updated the code per review. Changed to use `Either[String, A]` and the generated code for `hello_service` is this:

```scala
package ss.be.rpc.protos

import _root_.cats.implicits._

trait Greeter[F[_], A] {
  def sayHelloV1(request: ss.be.rpc.protos.HelloRequest, ctx: A): F[ss.be.rpc.protos.HelloReply]
  def sayHelloV2(request: ss.be.rpc.protos.HelloRequest, ctx: A): _root_.fs2.Stream[F, ss.be.rpc.protos.HelloReply]
  def sayHelloV3(request: _root_.fs2.Stream[F, ss.be.rpc.protos.HelloRequest], ctx: A): F[ss.be.rpc.protos.HelloReply]
  def sayHelloV4(request: _root_.fs2.Stream[F, ss.be.rpc.protos.HelloRequest], ctx: A): _root_.fs2.Stream[F, ss.be.rpc.protos.HelloReply]
}

object Greeter {
  
  def stub[F[_]: _root_.cats.effect.ConcurrentEffect](channel: _root_.io.grpc.Channel, callOptions: _root_.io.grpc.CallOptions = _root_.io.grpc.CallOptions.DEFAULT): Greeter[F, _root_.io.grpc.Metadata] = {
    client[F, _root_.io.grpc.Metadata](channel, identity, _ => callOptions)
  }
  
  def client[F[_]: _root_.cats.effect.ConcurrentEffect, A](channel: _root_.io.grpc.Channel, f: A => _root_.io.grpc.Metadata, c: _root_.io.grpc.CallOptions => _root_.io.grpc.CallOptions = identity): Greeter[F, A] = new Greeter[F, A] {
    def sayHelloV1(request: ss.be.rpc.protos.HelloRequest, ctx: A): F[ss.be.rpc.protos.HelloReply] = {
      _root_.org.lyranthe.fs2_grpc.java_runtime.client.Fs2ClientCall[F](channel, _root_.ss.be.rpc.protos.GreeterGrpc.METHOD_SAY_HELLO_V1, c(_root_.io.grpc.CallOptions.DEFAULT)).flatMap(_.unaryToUnaryCall(request, f(ctx)))
    }
    def sayHelloV2(request: ss.be.rpc.protos.HelloRequest, ctx: A): _root_.fs2.Stream[F, ss.be.rpc.protos.HelloReply] = {
      _root_.fs2.Stream.eval(_root_.org.lyranthe.fs2_grpc.java_runtime.client.Fs2ClientCall[F](channel, _root_.ss.be.rpc.protos.GreeterGrpc.METHOD_SAY_HELLO_V2, c(_root_.io.grpc.CallOptions.DEFAULT))).flatMap(_.unaryToStreamingCall(request, f(ctx)))
    }
    def sayHelloV3(request: _root_.fs2.Stream[F, ss.be.rpc.protos.HelloRequest], ctx: A): F[ss.be.rpc.protos.HelloReply] = {
      _root_.org.lyranthe.fs2_grpc.java_runtime.client.Fs2ClientCall[F](channel, _root_.ss.be.rpc.protos.GreeterGrpc.METHOD_SAY_HELLO_V3, c(_root_.io.grpc.CallOptions.DEFAULT)).flatMap(_.streamingToUnaryCall(request, f(ctx)))
    }
    def sayHelloV4(request: _root_.fs2.Stream[F, ss.be.rpc.protos.HelloRequest], ctx: A): _root_.fs2.Stream[F, ss.be.rpc.protos.HelloReply] = {
      _root_.fs2.Stream.eval(_root_.org.lyranthe.fs2_grpc.java_runtime.client.Fs2ClientCall[F](channel, _root_.ss.be.rpc.protos.GreeterGrpc.METHOD_SAY_HELLO_V4, c(_root_.io.grpc.CallOptions.DEFAULT))).flatMap(_.streamingToStreamingCall(request, f(ctx)))
    }
  }
  
  def bindService[F[_]: _root_.cats.effect.ConcurrentEffect](serviceImpl: Greeter[F, _root_.io.grpc.Metadata]): _root_.io.grpc.ServerServiceDefinition = {
    service[F, _root_.io.grpc.Metadata](serviceImpl, _.asRight[String])
  }
  
  def service[F[_]: _root_.cats.effect.ConcurrentEffect, A](serviceImpl: Greeter[F, A], f: _root_.io.grpc.Metadata => Either[String, A]): _root_.io.grpc.ServerServiceDefinition = {
    
    val g: _root_.io.grpc.Metadata ⇒ F[A] = f(_).leftMap[Throwable](_root_.io.grpc.Status.FAILED_PRECONDITION.withDescription(_).asRuntimeException()).raiseOrPure[F]
    
    _root_.io.grpc.ServerServiceDefinition
      .builder(_root_.ss.be.rpc.protos.GreeterGrpc.SERVICE)
      .addMethod(_root_.ss.be.rpc.protos.GreeterGrpc.METHOD_SAY_HELLO_V1, _root_.org.lyranthe.fs2_grpc.java_runtime.server.Fs2ServerCallHandler[F].unaryToUnaryCall[ss.be.rpc.protos.HelloRequest, ss.be.rpc.protos.HelloReply]((r, m) => g(m).flatMap(serviceImpl.sayHelloV1(r, _))))
      .addMethod(_root_.ss.be.rpc.protos.GreeterGrpc.METHOD_SAY_HELLO_V2, _root_.org.lyranthe.fs2_grpc.java_runtime.server.Fs2ServerCallHandler[F].unaryToStreamingCall[ss.be.rpc.protos.HelloRequest, ss.be.rpc.protos.HelloReply]((r, m) => _root_.fs2.Stream.eval(g(m)).flatMap(serviceImpl.sayHelloV2(r, _))))
      .addMethod(_root_.ss.be.rpc.protos.GreeterGrpc.METHOD_SAY_HELLO_V3, _root_.org.lyranthe.fs2_grpc.java_runtime.server.Fs2ServerCallHandler[F].streamingToUnaryCall[ss.be.rpc.protos.HelloRequest, ss.be.rpc.protos.HelloReply]((r, m) => g(m).flatMap(serviceImpl.sayHelloV3(r, _))))
      .addMethod(_root_.ss.be.rpc.protos.GreeterGrpc.METHOD_SAY_HELLO_V4, _root_.org.lyranthe.fs2_grpc.java_runtime.server.Fs2ServerCallHandler[F].streamingToStreamingCall[ss.be.rpc.protos.HelloRequest, ss.be.rpc.protos.HelloReply]((r, m) => _root_.fs2.Stream.eval(g(m)).flatMap(serviceImpl.sayHelloV4(r, _))))
      .build()
  }
}
```

rossabaker commented:
One of my colleagues suggested `ValidatedNel[String, A]`, in case we want to accumulate errors.  On the other hand, it's a slightly more exotic type.  I think it's a good idea, but could be easily persuaded to leave it as is.  What do you think?

Otherwise, this looks ready to me.

ahjohannessen commented:
@rossabaker I think `Either[String, A]` is ok in this case. The error ends up in `withDescription(_)` anyways. The user has the ability to decide how (s)he goes from `ValidatedNel[String, A]`, if used, to `Either[String, A]`, e.g. `mkString("\n")`, `reduce` or similar.

ahjohannessen commented:
@rossabaker @fiadliel What more is required in order to get this merged?

ahjohannessen commented:
@rossabaker 

>I tried to resolve the build.properties myself, but GitHub wouldn't let me. Once that's resolved, I'll merge and release.

I have rebased on master and updated the PR. Should be good to merge.

ahjohannessen commented:
👍
